Add Promise#finally proposal Stage 2 TC39 0716

clavecoder opened the pull request:
## [Promise#finally proposal Stage 2 from TC39 meeting on 2016-07](https://github.com/rwaldron/tc39-notes/blob/master/es7/2016-07/jul-28.md#10iiib-promiseprototypefinally).
- [proposal](https://github.com/rwaldron/tc39-notes/blob/master/es7/2016-07/jul-28.md#10iiib-promiseprototypefinally)
- [polyfill](https://github.com/tc39/proposal-promise-finally/blob/master/polyfill.js)

Used implementation of `String#padStart`, `String#trimLeft`, and `Promise#then` as examples. 
## Initial open items
- [X] Not able to use `finally` for the function name, perhaps because it is a reserved keyword. Temporarily used `_finally` for implementation and test (commented below).
- [X] Did not use polyfill's method of setting species. Coded to assume `core-js` Promise polyfill will set species, but, if I'm not mistaken, it's possible that `finally` may be `require`d without the Promise polyfill. Since `core-js` has its own method of setting species, what should be done?
- [ ] The failing test is precisely about Promise not having the right species (I think), but I'm not sure why, it shows up in the commonjs test for `delay`
  ![image](https://cloud.githubusercontent.com/assets/3524633/18235332/6136d86e-72df-11e6-98cf-71a51be870be.png)
- [ ] The `finally` implementation is not in an internal file, but directly in the imported one. I am not supposing that it would be used by other functions in the library. Should I rather assume that it may be?

clavecoder commented on library/modules/es7.promise.finally.js:
```diff
@@ -0,0 +1,27 @@
+var $export = require('./_export');
+
+$export($export.P, 'Promise', {
+  'finally': function _finally(onFinally){
```
Used `_finally`. Error generated without the underscore.

zloirock replied:
Use `'finally': function(onFinally)`. For ES6+ engines it will create the correct `.name`, for old engines we can't do anything.

clavecoder replied:
> Use 'finally': function(onFinally)

Done (not checked in yet...)

clavecoder replied:
Checked in.

clavecoder commented on modules/es7.promise.finally.js:
```diff
@@ -0,0 +1,27 @@
+var $export = require('./_export');
+
+$export($export.P, 'Promise', {
+  'finally': function _finally(onFinally){
```
Used `_finally`. Error generated without the underscore.

Kovensky replied:
```js
{
  finally (onFinally) {
    ...
  }
}
```

zloirock replied:
@Kovensky we haven't method shorthands in ES3 / ES5.

clavecoder commented on tests/tests/es7.promise.finally.ls:
```diff
@@ -0,0 +1,29 @@
+test 'Promise#finally' (assert)!->
+  assert.isFunction Promise::finally
+  NATIVE and assert.arity Promise::finally, 2 # FF - 0
+  assert.name Promise::finally, \_finally
```
Remove underscore when name is fixed to `finally` from `_finally`.

zloirock replied:
Remove this test.

clavecoder replied:
Done (not checked in  yet...).

clavecoder replied:
Checked in.

Kovensky commented on modules/es7.promise.finally.js:
```diff
@@ -0,0 +1,27 @@
+    var newPromise = Promise.prototype.then.call(
+      this, // throw if IsPromise(this) is not true
+      function(x) {
+        return new C()(function(resolve) {
```
You're calling the result of the `new C()` constructor as a function o_O

clavecoder replied:
That's the behavior of the polyfill:
https://github.com/tc39/proposal-promise-finally/blob/84cec8005364a0cae413cad3be1e47b6a26be369/polyfill.js#L33

Kovensky replied:
It crashes at runtime 🙀 

I've asked the author about it, but it's most likely accidental leftovers from refactoring.

I am using a previous version of the polyfill that had a much simpler `speciesConstructor` implementation (https://github.com/tc39/proposal-promise-finally/blob/c626fb5857e0564eff6ab6a5b5e2b9383c156ba7/polyfill.js).

zloirock replied:
So maybe it's a bug of this polyfill? It can't work.

clavecoder replied:
@kovensky:

1. Thanks for the older implementation reference. That's what I would expect. I like to move methodically through uncharted waters.
1. I'll use the core-js internal implementation of `speciesConstructor` used in `es6.promise.js`
1. Thanks for the quick responses!

Kovensky replied:
It behaves correctly without the excess `()`.

clavecoder replied:
Checked in.

Kovensky commented on modules/es7.promise.finally.js:
```diff
@@ -0,0 +1,27 @@
+$export($export.P, 'Promise', {
+  'finally': function _finally(onFinally){
+    var handler = typeof onFinally === 'function' ? onFinally : function(){};
+    var C = Promise;
```
You're hardcoding the constructor as Promise instead of the species constructor

zloirock replied:
Use `_species-constructor`. Also, what is `Promise`? We can't use this global variable, if you need it for something - you can use `core.Promise || global.Promise`.

clavecoder replied:
That's why my second open item is asking the right way in core-js to get the species constructor. I think I've found it now (not checked in yet) but I'm still getting `TypeError` with `delay` (see image in initial comment).

clavecoder replied:
Checked in.

zloirock commented on fn/promise/index.js:
```diff
@@ -0,0 +1,2 @@
+require('../../modules/es7.promise.finally');
+module.exports = require('../../modules/_core').Promise;
```
We already have `fn/promise.js`. Also, here missed `es6.promise`.

clavecoder replied:
> We already have `fn/promise.js`
OK, `require`d es7.promise.finally in `fn/promise.js` in leu of `fn/promise/finally.js`. Or do you want them both, so they can just add finally?

clavecoder replied:
Checked in. Note that `fn/promise/` has been removed.

zloirock commented on tests/tests/es7.promise.finally.ls:
```diff
@@ -0,0 +1,29 @@
+  FakePromise1[Symbol?species] = !-> it ->, null
+  assert.throws (!-> promise.finally(->)), 'NewPromiseCapability validations, #3'
+
+
```
Too many empty lines. Also, where `tests/library/es7.promise.finally.ls`?

clavecoder replied:
Extra lines removed and `tests/library/es7.promise.finally.ls` added with tests copied from `then` (not checked in yet). Working on real tests...

clavecoder replied:
Checked in. Only have pseudocode. :frowning: Can you point me to a similar example?

zloirock commented on tests/tests/es7.promise.finally.ls:
```diff
@@ -0,0 +1,29 @@
+
+test 'Promise#finally' (assert)!->
+  assert.isFunction Promise::finally
+  NATIVE and assert.arity Promise::finally, 2 # FF - 0
```
`FF - 0`?

clavecoder replied:
Left over from cut/paste of `then`. Removed (not checked in yet).

clavecoder replied:
Checked in.

zloirock commented on tests/tests/es7.promise.finally.ls:
```diff
@@ -0,0 +1,29 @@
+  FakePromise1[Symbol?species] = !-> it null, ->
+  assert.throws (!-> promise.finally(->)), 'NewPromiseCapability validations, #2'
+  FakePromise1[Symbol?species] = !-> it ->, null
+  assert.throws (!-> promise.finally(->)), 'NewPromiseCapability validations, #3'
```
I see only subclassing tests. For `Promise#then` it's enough because of the main test case for them - https://github.com/promises-aplus/promises-tests, but `finally` missed here.

clavecoder replied:
Checked in. Only have pseudocode. :frowning: Can you point me to a similar example?

clavecoder commented:
Checked in changes per comments. However my added test file is failing hard. Changes made:

1. `es7/promise.js`: Added require of `es6.promise` before `es7.promise.finally`
1. `fn/promise.js`: Added require of `es7.promise.finally` after `es6.promise`
1. Removed `fn/promise/` in favor of `fn/promise.js`
1. `library/es7/promise.js`: Added require of `es6.promise` before `es7.promise.finally`
1. `library/modules/es7.promise.finally.js`:
   1. changed to implicit name for `finally`
   1. now uses `speciesConstructor`
   1. Removed polyfill regression of using Promise instance as function.
1. `modules/es7.promise.finally.js`: *ditto*
1. Added `/tests/library/es7.promise.finally.ls`
   1. Mirrored behavior of `/tests/tests/es7.promise.finally.ls`
1. `/tests/tests/es7.promise.finally.ls`
   1. Removed extraneous comment
   1. Removed test for name of Promise#finally
   1. Added JavaScript pseudocode for Promise#finally implementation tests.

Open issues:
- [ ] `tests/library.js` is failing because `Promise#finally` is not found for test file `/tests/library/es7.promise.finally.ls`. I've tried tracing the `require` path and, of course, it seems right, but there's obviously something wrong. Also remember that the implementation was changed to use an anonymous function for the property assignment. I don't know if it is related. Error message:
![image](https://cloud.githubusercontent.com/assets/3524633/18257290/c65f3afe-7386-11e6-9fa0-d91306c136a6.png)
The source code for the error line is `/tests/library/es7.promise.finally.ls/L7`
```
  assert.isFunction Promise::finally
```
- [ ] I'm finding LiveScript a bit beyond me for creating the unit tests without an example. I've added JavaScript pseudocode in comments  for `/tests/library/es7.promise.finally.ls` and `/tests/tests/es7.promise.finally.ls`. If you can point me to some LiveScript tests that are doing something similar, I should be able to tease out the implementation.

clavecoder commented:
Any test code you can point me to that work similarly to the pseudocode? I should have more time this weekend.

cvbuelow commented:
Is this still being worked on?

zloirock commented:
https://github.com/zloirock/core-js/commit/ed59a9c0822eb1fa3509e06f996f424d8392d0a9, https://github.com/zloirock/core-js/commit/86ed553fbaa2d1e2f380bfe174f158e711e2e51f, https://github.com/zloirock/core-js/commit/fd70e24f8f076177c321b29aa2dc3b4de8592122, https://github.com/zloirock/core-js/commit/0bc7a9ea75fd1a604cc6c99eb755c32a08afda9a
